packages/integration-tests/src/test-e2e/amplify_outputs_backwards_compatibility.test.ts

```diff
@@ -0,0 +1,188 @@
+    await cfnClient.send(
+      new DeleteStackCommand({
+        StackName: BackendIdentifierConversions.toStackName(
+          branchBackendIdentifier
+        ),
+      })
+    );
+    await fsp.rm(tempDir, { recursive: true });
+
+    await baselineNpmProxyController.tearDown();
+    await currentNpmProxyController.tearDown();
+  });
+
+  const deploy = async (): Promise<void> => {
+    await execa(
+      'npx',
+      [
+        'ampx',
+        'pipeline-deploy',
+        '--branch',
+        branchBackendIdentifier.name,
+        '--appId',
+        branchBackendIdentifier.namespace,
+      ],
+      {
+        cwd: tempDir,
+        stdio: 'inherit',
+        env: {
+          CI: 'true',
+        },
+      }
+    );
+  };
+
+  const reinstallDependencies = async (): Promise<void> => {
+    await fsp.rm(path.join(tempDir, 'node_modules'), {
+      recursive: true,
+      force: true,
+    });
+    await fsp.unlink(path.join(tempDir, 'package-lock.json'));
+
+    await execa('npm', ['install'], {
+      cwd: tempDir,
+      stdio: 'inherit',
+    });
+  };
+
+  void it('outputs generation should be backwards and forward compatible', async () => {
+    // build an app using previous (baseline) version
+    await baselineNpmProxyController.setUp();
+    await execa('npm', ['create', amplifyAtTag, '--yes'], {
+      cwd: tempDir,
+      stdio: 'inherit',
+    });
+
+    // Replace backend.ts to add custom outputs without version as well.
+    await fsp.writeFile(
+      path.join(tempDir, 'amplify', 'backend.ts'),
+      `import { defineBackend } from '@aws-amplify/backend';
+import { auth } from './auth/resource';
+import { data } from './data/resource';
+
+const backend = defineBackend({
+  auth,
+  data,
+});
+
+backend.addOutput({
+  custom: {
+    someCustomOutput: 'someCustomOutputValue',
+  },
+});
+`
+    );
+    await deploy();
+    await baselineNpmProxyController.tearDown();
+
+    // Generate the outputs using the new version
+    await currentNpmProxyController.setUp();
+    await reinstallDependencies();
+
+    await execa(
+      'npx',
+      [
+        'ampx',
+        'generate',
+        'outputs',
+        '--stack',
+        BackendIdentifierConversions.toStackName(branchBackendIdentifier),
+      ],
+      {
+        cwd: tempDir,
+        stdio: 'inherit',
+      }
+    );
+    let fileSize = (await fsp.stat(path.join(tempDir, 'amplify_outputs.json')))
+      .size;
+    assert.ok(
+      fileSize > 100, // Validate that it's not just a shim
+      'outputs file should not be empty when generating for an older app with new version'
+    );
+    // delete the file now
+    await fsp.unlink(path.join(tempDir, 'amplify_outputs.json'));
+
+    // Re-deploy the app using the new version now
+    await deploy();
+
+    // Generate the outputs using the older version
+    await currentNpmProxyController.tearDown();
+    await baselineNpmProxyController.setUp();
+    await reinstallDependencies();
+    await execa(
```

Comment:
Should we also add code that calls `generateClientConfig<>` with all known versions ?

Reply:
Added